README.md

```diff
@@ -1 +1,78 @@
 # rancher-desktop-networking
+`rancher-desktop-networking` is a simple layer 2 switch that allows connectivity between the VM and the host over `AF_VSOCK` written in Go using [gvisor's](https://github.com/google/gvisor) network stack and [gvisor-tap-vsock](https://github.com/google/gvisor). In addition to piping the ethernet frames, it also provides a DNS, DHCP services and dynamic port forwarding.
+
+Rancher Desktop Networking consists of three main components: host-switch, vm-switch, and network-setup. Below is a brief explanation of what each component does.
+
+## How it works
+```mermaid
+flowchart  LR;
+ subgraph Host["HOST"]
+ subgraph hostSwitch["host-switch.exe"]
+ vsockHost{"main loop"}
+ eth(("reconstruct ETH frames"))
+ syscall(("OS syscall"))
+ dhcp["DHCP"]
+ dns["DNS"]
+ api["API"]
+portForwarding["Port Forwarding"]
+ vsockHost <----> eth
+ eth <----> syscall
+ vsockHost ----> dhcp
+ vsockHost ----> dns
+ vsockHost ----> api
+ vsockHost ----> portForwarding
+ end
+ end
+ subgraph VM["VM"]
+ subgraph netNs["Isolated Network Namespace"]
+ tapDevice("eth0")
+ subgraph vmSwitch["vm-switch"]
+ vsockVM{"VM Switch"}
+ ethVM(("listens for ETH frames\n from TAP Device"))
+ ethVM <----> vsockVM
+ end
+ tapDevice <----> ethVM
+ init["/sbin/init"]
+ end
+ network-setup["/usr/local/bin/network-setup"]
+ end
+ vsockVM  <---> |AF_VSOCK| vsockHost
+```
+
+## host-switch
+`host-switch` runs on the Windows host and acts as a receiver for all traffic originating from the network namespace within the WSL VM. It performs a handshake to find the right VM to talk to over `AF_VSOCK`. Once a correct VM is found, it then listens for the incoming traffic from that VM. In additon to this, it can provide a DNS resolver that runs in the user space network along with an API that allows for dynamic port forwarding.
+
+## network-setup
+Its main responsibility is to respond to the handshake request from the `host-switch.exe`, create a network namespace and start the `vm-switch` subprocess in the newly created network namespace. Below is a sequence diagram demonstrating the process.
+
+```mermaid
+sequenceDiagram
+    box vm
+    participant script as wsl-init (pid n)
```

Comment:
It looks like GitHub's mermaid renderer doesn't like something here… Maybe it doesn't like `box`? https://github.com/mermaid-js/mermaid/issues/4098 indicates it might be relatively new.